File: festival-kotlin-web/src/main/java/cn/bdqfork/web/server/AbstractWebServer.java
package cn.bdqfork.web.server;

import cn.bdqfork.core.exception.BeansException;
import cn.bdqfork.web.RouterAware;
import cn.bdqfork.web.VertxAware;
import io.vertx.core.Vertx;
import io.vertx.ext.web.Router;

/**
 * @author bdq
 * @since 2020/2/12
 */
public abstract class AbstractWebServer implements WebServer, RouterAware, VertxAware {
    protected Vertx vertx;
    protected Router router;
    private boolean running;

    @Override
    public void start() throws Exception {

        registerCoreHandler(router);

        registerOptionHandler(router);

        registerRouteMapping(router);

        doStart();
    }

    protected abstract void registerRouteMapping(Router router) throws Exception;

    protected abstract void registerOptionHandler(Router router) throws Exception;

    protected abstract void registerCoreHandler(Router router) throws Exception;

    protected abstract void doStart() throws Exception;


    @Override
    public void stop() throws Exception {
        if (running) {
            doStop();
            running = false;
        }
    }

    protected abstract void doStop() throws Exception;

    public boolean isRunning() {
        return running;
    }

    @Override
    public void setRouter(Router router) throws BeansException {
        this.router = router;
    }

    @Override
    public void setVertx(Vertx vertx) throws BeansException {
        this.vertx = vertx;
    }
}

File: festival-kotlin-web/src/main/java/cn/bdqfork/web/server/AbstractWebServer.kt
package cn.bdqfork.web.server

import cn.bdqfork.core.exception.BeansException
import cn.bdqfork.web.RouterAware
import cn.bdqfork.web.VertxAware
import io.vertx.core.Vertx
import io.vertx.ext.web.Router

/**
 * @author bdq
 * @since 2020/2/12
 */
abstract class AbstractWebServer : WebServer, RouterAware, VertxAware {
    @JvmField
    protected var vertx: Vertx? = null
    @JvmField
    protected var router: Router? = null
    private var isRunning = false

    @Throws(Exception::class)
    override fun start() {
        registerCoreHandler(router)
        registerOptionHandler(router)
        registerRouteMapping(router)
        doStart()
    }

    @Throws(Exception::class)
    protected abstract fun registerRouteMapping(router: Router?)

    @Throws(Exception::class)
    protected abstract fun registerOptionHandler(router: Router?)

    @Throws(Exception::class)
    protected abstract fun registerCoreHandler(router: Router?)

    @Throws(Exception::class)
    protected abstract fun doStart()

    @Throws(Exception::class)
    override fun stop() {
        if (isRunning) {
            doStop()
            isRunning = false
        }
    }

    @Throws(Exception::class)
    protected abstract fun doStop()

    @Throws(BeansException::class)
    override fun setRouter(router: Router?) {
        this.router = router
    }

    @Throws(BeansException::class)
    override fun setVertx(vertx: Vertx) {
        this.vertx = vertx
    }
}